Zend/zend_object_handlers.c

```diff
@@ -839,6 +828,16 @@ ZEND_API zval *zend_std_write_property(zend_object *zobj, zend_string *name, zva
 					variable_ptr = &EG(error_zval);
 					goto exit;
 				}
+				if (UNEXPECTED(prop_info->flags & ZEND_ACC_READONLY)) {
+					if (Z_PROP_FLAG_P(variable_ptr) & IS_PROP_REINITABLE) {
+						Z_PROP_FLAG_P(variable_ptr) &= ~IS_PROP_REINITABLE;
+					} else {
+						Z_TRY_DELREF_P(value);
```

Comment:
I *think* this should release `&tmp` instead. If `zend_verify_property_type` coerces the type it can allocate a new value (e.g. int -> string).

```c
zval_ptr_dtor(&tmp);
```

Otherwise:

```php
--TEST--
Test failing readonly assignment with coercion
--FILE--
<?php

class Foo {
    public readonly string $bar;

    public function __construct() {
        $this->bar = 'bar';
        try {
            $this->bar = 42;
        } catch (Error $e) {
            echo $e->getMessage(), "\n";
        }
    }
}

new Foo();

?>
--EXPECTF--
Cannot modify readonly property Foo::$bar

=================================================================
==335637==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 32 byte(s) in 1 object(s) allocated from:
    #0 0x7f65882ba6af in __interceptor_malloc (/lib64/libasan.so.8+0xba6af)
    #1 0x1527c1e in tracked_malloc /home/ilutov/Developer/php-src/Zend/zend_alloc.c:2826
    #2 0x1525fbd in _malloc_custom /home/ilutov/Developer/php-src/Zend/zend_alloc.c:2471
    #3 0x1526389 in _emalloc /home/ilutov/Developer/php-src/Zend/zend_alloc.c:2590
    #4 0x15f2a4b in zend_string_alloc /home/ilutov/Developer/php-src/Zend/zend_string.h:172
    #5 0x15f2c58 in zend_string_init /home/ilutov/Developer/php-src/Zend/zend_string.h:194
    #6 0x1621764 in zend_long_to_str /home/ilutov/Developer/php-src/Zend/zend_operators.c:3249
    #7 0x15fbd63 in _convert_to_string /home/ilutov/Developer/php-src/Zend/zend_operators.c:715
    #8 0x1648325 in zend_parse_arg_str_weak /home/ilutov/Developer/php-src/Zend/zend_API.c:693
    #9 0x1709a54 in zend_verify_weak_scalar_type_hint /home/ilutov/Developer/php-src/Zend/zend_execute.c:761
    #10 0x170a0bc in zend_verify_scalar_type_hint /home/ilutov/Developer/php-src/Zend/zend_execute.c:828
    #11 0x170c115 in i_zend_check_property_type /home/ilutov/Developer/php-src/Zend/zend_execute.c:991
    #12 0x170c147 in i_zend_verify_property_type /home/ilutov/Developer/php-src/Zend/zend_execute.c:996
    #13 0x170c198 in zend_verify_property_type /home/ilutov/Developer/php-src/Zend/zend_execute.c:1005
    #14 0x1abe8d1 in zend_std_write_property /home/ilutov/Developer/php-src/Zend/zend_object_handlers.c:818
    #15 0x18f5ffe in ZEND_ASSIGN_OBJ_SPEC_UNUSED_CONST_OP_DATA_CONST_HANDLER /home/ilutov/Developer/php-src/Zend/zend_vm_execute.h:33004
    #16 0x19ebdd8 in execute_ex /home/ilutov/Developer/php-src/Zend/zend_vm_execute.h:59452
    #17 0x19f504f in zend_execute /home/ilutov/Developer/php-src/Zend/zend_vm_execute.h:60935
    #18 0x163a866 in zend_execute_scripts /home/ilutov/Developer/php-src/Zend/zend.c:1798
    #19 0x13cfd7e in php_execute_script /home/ilutov/Developer/php-src/main/main.c:2482
    #20 0x1d86225 in do_cli /home/ilutov/Developer/php-src/sapi/cli/php_cli.c:964
    #21 0x1d8888d in main %s:%d
    #22 0x7f658764a50f in __libc_start_call_main (/lib64/libc.so.6+0x2750f)

SUMMARY: AddressSanitizer: 32 byte(s) leaked in 1 allocation(s).
```

Reply:
Thanks again! I'vd just fixed this issue and added the test from your comment!